pkg/agent/controller/ipseccertificate/ipsec_certificate_controller.go

```diff
@@ -0,0 +1,458 @@
+		}
+	}
+	if pair.privateKeyPath != "" {
+		// Delete the old private key file.
+		if err := os.Remove(pair.privateKeyPath); err != nil && !os.IsNotExist(err) {
+			klog.ErrorS(err, "Failed to delete old private key", "file", pair.privateKeyPath)
+		}
+	}
+}
+
+// jitteryDuration returns a duration in [totalDuration * 0.7, totalDuration * 0.9].
+var jitteryDuration = func(totalDuration time.Duration) time.Duration {
+	// wait.Jitter returns a duration in [totalDuration, totalDuration * 1.2].
+	return wait.Jitter(time.Duration(totalDuration), 0.2) - time.Duration(float64(totalDuration)*0.3)
+}
+
+// nextRotationDeadline returns a value for the threshold at which the
+// current certificate should be rotated, 80%+/-10% of the expiration of the
+// certificate. The deadline will not change once calculated.
+// This function is not thread-safe.
+func (pair *certificateKeyPair) nextRotationDeadline() time.Time {
+	if pair == nil || len(pair.certificate) == 0 {
+		return time.Now()
+	}
+	// Return the previous calculated rotation deadline if applicable.
+	if !pair.rotationDeadline.IsZero() {
+		return pair.rotationDeadline
+	}
+	notAfter := pair.certificate[0].NotAfter
+	totalDuration := notAfter.Sub(pair.certificate[0].NotBefore)
+	deadline := pair.certificate[0].NotBefore.Add(jitteryDuration(totalDuration))
+	klog.InfoS("Calculated certificate rotation deadline", "expiration", notAfter, "deadline", deadline)
+	pair.rotationDeadline = deadline
+	return deadline
+}
+
+func loadCertAndKeyFromFiles(caPath, certPath, keyPath string) (*certificateKeyPair, error) {
+	ca, err := loadRootCA(caPath)
+	if err != nil {
+		return nil, err
+	}
+	key, err := loadPrivateKey(keyPath)
+	if err != nil {
+		return nil, err
+	}
+	cert, err := loadCertificate(certPath)
+	if err != nil {
+		return nil, err
+	}
+	pair := &certificateKeyPair{
+		certificatePath: certPath,
+		privateKeyPath:  keyPath,
+		caCertificate:   ca,
+		certificate:     cert,
+		privateKey:      key,
+	}
+	return pair, nil
+}
+
+func (c *Controller) syncConfigurations() error {
+	startTime := time.Now()
+	defer func() {
+		d := time.Since(startTime)
+		klog.V(2).InfoS("Finished syncing IPsec certificate configurations", "duration", d)
+	}()
+
+	var deadline time.Time
+	// Validate the existing certificate and key pair.
+	if err := c.certificateKeyPair.validate(); err != nil {
+		klog.ErrorS(err, "Verifying current certificate configurations failed")
+		deadline = time.Now()
+	} else {
+		deadline = c.certificateKeyPair.nextRotationDeadline()
+	}
+	// Current certificate is about to expire.
+	if sleepInterval := time.Until(deadline); sleepInterval <= 0 {
+		klog.InfoS("Start rotating IPsec certificate")
+		newCertPath, newKeyPath, err := c.rotateCertificate()
+		if err != nil {
+			return fmt.Errorf("failed to rotate certificate: %w", err)
+		}
+		if c.certificateKeyPair != nil {
+			c.certificateKeyPair.cleanup()
+		}
+		newPair, err := loadCertAndKeyFromFiles(c.caPath, newCertPath, newKeyPath)
```

Comment:
why it needs to load certificate from files after it just generates it? couldn't `c.rotateCertificate` just return a new certificate pair?

Reply:
maybe better to name the function `c.newCertificateKeyPair`

Reply:
Originally the sync loop loads files for each run. It is not required now. Changed the rotateCertificate function to return certificate pair as suggested.